src/v/iceberg/table_update_applier.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/table_update_applier.h"
+
+#include "base/vlog.h"
+#include "container/chunked_hash_map.h"
+#include "container/fragmented_vector.h"
+#include "iceberg/logger.h"
+#include "iceberg/schema.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/table_update.h"
+
+namespace iceberg::table_update {
+
+namespace {
+
+struct update_applying_visitor {
+    explicit update_applying_visitor(table_metadata& meta)
+      : meta(meta) {}
+    table_metadata& meta;
+
+    outcome operator()(const add_schema& update) {
+        auto sid = update.schema.schema_id;
+        auto s = std::find_if(
+          meta.schemas.begin(),
+          meta.schemas.end(),
+          [sid](const schema& schema) { return schema.schema_id == sid; });
```

Comment:
Not a nit or suggestion but this becomes quite concise with `std::ranges` and projection:
```
auto s = std::ranges::find(meta.schemas, sid, &schema::schema_id);
```

Reply:
Thanks for the tip! This is very nice. Will update in a follow-up